src/Middleware/HeaderPropagation/src/HeaderPropagationMiddleware.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    public class HeaderPropagationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HeaderPropagationOptions _options;
+        private readonly HeaderPropagationState _state;
+
+        public HeaderPropagationMiddleware(RequestDelegate next, IOptions<HeaderPropagationOptions> options, HeaderPropagationState state)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _options = options.Value;
+
+            _state = state ?? throw new ArgumentNullException(nameof(state));
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            if (context != null)
+            {
+                foreach (var header in _options.Headers)
+                {
+                    if (!context.Request.Headers.TryGetValue(header.InputName, out var values)
+                        || StringValues.IsNullOrEmpty(values))
+                    {
+                        if (header.DefaultValuesGenerator != null)
+                        {
+                            values = header.DefaultValuesGenerator(context);
+                            if (StringValues.IsNullOrEmpty(values)) continue;
+                        }
+                        else if (!StringValues.IsNullOrEmpty(header.DefaultValues))
+                        {
+                            values = header.DefaultValues;
+                        }
+                        else
+                        {
+                            continue;
+                        }
+                    }
+
+                    var outputName = !string.IsNullOrEmpty(header.OutputName) ? header.OutputName : header.InputName;
```

Comment:
It doesn't look like the message handler handles this case. It will try to use `OutputName` regardless.

Reply:
You are right, I fixed it doing a change:
I removed the logic in the middleware and saved in the "state" using the input name, then moved the logic in the handler.
This is because to address the ability to specify the headers to be propagated for each client, there will be a part of config needed for the `DelegatingHandler` and I believe it will be the right place to also specify the output name (so calls to different services can use different header names).